Add 'axis' option for 'tf.boolean_mask()'

yongtang opened the pull request:
This fix tries to address #9721 where it was not possible to pass an
'axis' option for 'tf.boolean_mask()'.

This fix fixes #9721.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add 'axis' option for 'tf.boolean_mask()'

This fix tries to address 9721 where it was not possible to pass an
'axis' option for 'tf.boolean_mask()'.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add tests for 'axis' option in 'tf.boolean_mask()'

This fix adds tests for 'axis' option in 'tf.boolean_mask()'

This fix is part of the effort to fix 9721.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

tensorflow-jenkins commented:
Can one of the admins verify this patch?

josh11b commented on tensorflow/python/ops/array_ops.py:
```diff
@@ -1084,6 +1084,7 @@ def boolean_mask(tensor, mask, name="boolean_mask"):
   Args:
     tensor:  N-D tensor.
     mask:  K-D boolean tensor, K <= N and K must be known statically.
+    axis: The axis in `tensor` to mask from. Defaults to the first dimension.
```
What if mask isn't a 1-D tensor?

yongtang replied:
@josh11b If mask isn't 1-D then it should work as well. I added an additional test case for this.

josh11b commented on tensorflow/python/ops/array_ops.py:
```diff
@@ -1066,7 +1066,7 @@ def concat(values, axis, name="concat"):
 
 
-def boolean_mask(tensor, mask, name="boolean_mask"):
+def boolean_mask(tensor, mask, axis=None, name="boolean_mask"):
```
For backwards compatibility, all new arguments must be last (in addition to having defaults).

yongtang replied:
Thanks @josh11b. The PR has been updated with the suggested changes.

josh11b reviewed (commented):
(API review.)

yongtang committed:
Move the new args 'axis=None' to the end for backward compatibility

This fix move the new args 'axis=None' to the end for backward compatibility
and adds additional tests to cover cases where mask is more than 1-D.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@josh11b Thanks for the review. The PR has been updated. Please take a look.

josh11b commented on tensorflow/python/ops/array_ops.py:
```diff
@@ -1085,6 +1085,7 @@ def boolean_mask(tensor, mask, name="boolean_mask"):
     tensor:  N-D tensor.
     mask:  K-D boolean tensor, K <= N and K must be known statically.
     name:  A name for this operation (optional).
+    axis: The axis in `tensor` to mask from. Defaults to the first dimension.
```
You should clarify the documentation. In particular you should mention that it must be an integer, and please update lines 1080-1082. Ideally in addition `axis` would also be allowed to be a list of length K to be consistent with other `axis` arguments in the API, e.g. https://www.tensorflow.org/api_docs/python/tf/reduce_sum .

langmore replied:
`boolean_mask` already can work as an n-dimensional mask.  Currently if `mask.shape.ndims = K`, it masks over the first `K` dimensions of `tensor`.  So the docstring should be changed to reflect the new behavior if e.g. `axis != None` and `mask` is multi-dimensional.

langmore replied:
`boolean_mask(tensor, mask)` is supposed to act like numpy's `my_array[my_mask]`, which it currently does.  What you're proposing here looks like [`numpy.compress`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.compress.html#numpy.compress)

So it seems like this functionality should be part of an `array_ops.compress`.

josh11b replied:
So numpy.compress only allows 1-D boolean masks, which makes the choice of only using an integer for the axis straightforward.  This op allows K-D boolean masks, which raises the question of whether the axis should be (allowed to be) a length-K vector.

josh11b reviewed (dismissed):
(for API review)

jhseu commented:
@yongtang Any updates? Thanks!

yongtang committed:
Update docstring for 'tf.boolean_mask()'

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@josh11b @langmore Sorry for missing the GitHub notification and late reply. I have rebased and updated PR with the docstrings addressed based on the comments.

Please take a look and let me know if there are any issues.

drpngx commented:
@langmore @josh11b any additional thoughts?

Jenkins, test this please.

sb2nov commented:
@langmore @josh11b any additional thoughts?

josh11b commented:
This is good for api review.

yongtang committed:
Update goldens for api compatibility changes.

This commit update API goldens.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @josh11b for the review. The API goldens have been updated so that `//tensorflow/tools/api/tests:api_compatibility_test` could be passed.

drpngx commented:
Jenkins, test this please.

martinwicke commented:
Ignoring Kokoro -- looks like tool failure to me.
